controller/array_action/array_mediator_ds8k.py

```diff
@@ -370,17 +374,143 @@ def unmap_volume(self, volume_id, host_name):
+        volume_candidates = []
+
+        for pool in pools:
+            try:
+                volume_candidates.extend(self.client.get_volumes_by_pool(pool.id))
+            except exceptions.NotFound as ex:
+                if ERROR_CODE_RESOURCE_NOT_EXISTS in str(ex.message).upper():
+                    raise array_errors.PoolDoesNotExist(pool,
+                                                        self.identifier)
+                else:
+                    raise ex
+            for volume in volume_candidates:
+                logger.info("Checking volume: {}".format(volume.name))
+                if volume.name == shorten_volume_name(volume_name, prefix=''):
+                    logger.debug("Found volume: {}".format(volume))
+                    return volume
+        return None
+
+    def _get_volume_by_name(self, volume_name):
+        api_volume = self._get_api_volume_by_name(volume_name)
+        return self._generate_volume_response(api_volume)
+
+    def _get_api_volume_by_id(self, volume_id):
+        try:
+            return self.client.get_volume(volume_id)
+        except exceptions.NotFound:
+            raise array_errors.VolumeNotFoundError(volume_id)
+
+    def _get_flashcopy(self, fc_id, not_exist_err=True):
+        logger.info("Getting flashcopy {}".format(fc_id))
+        try:
+            return self.client.get_flashcopies(fc_id)
+        except exceptions.NotFound as ex:
+            if ERROR_CODE_RESOURCE_NOT_EXISTS in str(ex.message).upper():
+                logger.info("{} not found".format(fc_id))
+                if not_exist_err:
+                    raise ex
+        except Exception as ex:
+            logger.exception(ex)
+            raise ex
+
     def get_snapshot(self, snapshot_name):
-        # TODO: CSI-1339
-        pass
+        logger.debug("Get snapshot : {}".format(snapshot_name))
+        target_api_volume = self._get_api_volume_by_name(snapshot_name)
+        if not target_api_volume:
+            return None
+        if not target_api_volume.flashcopy:
+            logger.error(
+                "FlashCopy relationship not found for target volume: {}".format(snapshot_name))
+            raise array_errors.SnapshotNameBelongsToVolumeError(target_api_volume.name,
+                                                                self.service_address)
+        fcrel = self._get_flashcopy(target_api_volume.flashcopy[0].id)
+        source_volume_name = self.get_volume_name(fcrel.source_volume['id'])
+        return self._generate_snapshot_response(target_api_volume, source_volume_name)
+
+    def _create_similar_volume(self, target_volume_name, source_volume_name):
+        logger.info(
+            "creating target api volume '{0}' from source volume '{1}'".format(target_volume_name,
+                                                                               source_volume_name))
+        source_api_volume = self._get_api_volume_by_name(source_volume_name)
+        if not source_volume_name:
```

Comment:
[this bug](https://github.com/IBM/ibm-block-csi-driver/pull/165/files#diff-136d3e5df32555e5e0bcfcfb6496e862R456) should be fixed in this PR

Reply:
done